### ISSUE TITLE
Add support for array style options

### ISSUE THREAD

**Issue by tdulcet** [TinyWebEx/AutomaticSettings#15]
This may be out of scope for this library, but it would be very useful if it supported array style options, where the user can add more than one value to an option or more than one instance of an option group. For example, consider a standard weather app/extension where the user can add multiple locations, each with a ZIP code and country. This would also be needed for any inputs where the [`multiple` attribute](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#attr-multiple) is set.

**Comment by tdulcet** [comment]
> Apart from that, I don't consider it out-of-scope, it's just hard to implement technically and I never really needed it and could find other solutions, so it's not in there.

Yeah, it would require some significant JS to implement the UI.

> If you have an idea or want/need an implementation, I'd be happy to look at PRs and better maybe discuss how we should approach/handle it, before. 😃

I have at least two extension ideas which would need this feature, however I think the four other issues I created are a higher priority, especially the two blocking parts of Unicodify and the Awesome Emoji Picker, although I would definitely be interested to hear how you think we could best implement this feature...

Back to my weather extension example, I was thinking developers could specify their default settings like this to indicate that they want to use array style options for an option group:
```js
const defaultSettings = {
	locations: [
		{
			zip: "20500",
			country: "US"
		}
	]
};
```

**Comment by tdulcet** [comment]
> > Apart from that, I don't consider it out-of-scope, it's just hard to implement technically and I never really needed it and could find other solutions, so it's not in there.
> 
> Yeah, it could require some significant JS to implement the UI.

For adding more than one instance of an option group, I implemented a simple UI using a table, which has a column for each option in the option group and a row for each instance of the option group: https://jsfiddle.net/4bafm35n/. There is also a button to covert the table to an array, to demonstrate what would be needed by this library to save the values in the table. This library would then just need run that code anytime the user modified any of the cells.

For the select elements where the `mutiple` attribute is set, there is a one liner here to convert that to an array: https://stackoverflow.com/a/39363742, which is what would be needed for https://github.com/rugk/unicodify/issues/30.

> I'd be happy to look at PRs and better maybe discuss how we should approach/handle it, before. 😃

I would also be happy to discuss how we could best update the library to support this feature...

**Comment by rugk** [comment]
What you read sounds all goods far as for what you write. A table sounds reasonable as a HTML markup for that, though I maybe would use that in the code, but rather use classes, so you can use your own markup there (after all, maybe the options also are not supposed to be in one line/could spawn multiple lines).

So maybe it could be designed [similar to how the options groups are implemented](https://github.com/TinyWebEx/AutomaticSettings#option-groups).

Instead of a string as a thing to copy for each entry/"line"/"row", it should use a [HTML <template>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template). Maybe also a [slot](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot) for where this stuff, but I am not sure how this WebComponent thing is used.

Finally, it just needs a button with another class that is used as the "add another entry" button.
Maybe also a class for the "remove this entry" button that can be used in the `<template>` to delete that entry.
And finally maybe another class for "clear all".

**Comment by tdulcet** [comment]
> What you read sounds all goods far as for what you write. A table sounds reasonable as a HTML markup for that, though I maybe would use that in the code, but rather use classes, so you can use your own markup there (after all, maybe the options also are not supposed to be in one line/could spawn multiple lines).

Yes, anyone using your library would need to write their own HTML for the table. That link was just to demo a possible UI for this, so that we could make the necessary changes to the library. Specifically, it was using the weather extension example from my above post.

> Instead of a string as a thing to copy for each entry/"line"/"row", it should use a [HTML . Maybe also a ](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template)[slot](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/slot) for where this stuff, but I am not sure how this WebComponent thing is used.

Yes, I think the HTML Template element would work great here, as the user could then specify the format for each row directly in their HTML file.

> Finally, it just needs a button with another class that is used as the "add another entry" button. Maybe also a class for the "remove this entry" button that can be used in the `<template>` to delete that entry. And finally maybe another class for "clear all".

I am not sure what you mean by "class". A JS class? A CSS class? I updated my demo to support removing all the entries/rows, as well as moving individual rows up and down: https://jsfiddle.net/4bafm35n/1/. I also used emojis to reduce the amount of text that would need to be localized.

Here is another updated version using the HTML Template element you suggested: https://jsfiddle.net/4bafm35n/3/. This also now ignores any rows with invalid cells when creating the array, using the method suggested in #14.

**Comment by rugk** [comment]
> what you mean by "class". A JS class? A CSS class?

The CSS one.

> I also used emojis to reduce the amount of text that would need to be localized.

Ah indeed a good idea.

> ere is another updated version using the HTML Template element as you suggested: https://jsfiddle.net/4bafm35n/3/. This also now ignores any rows with invalid cells when creating the array, using the method suggested in #14.

Nice one.

**Comment by tdulcet** [comment]
Similar to #13, I think this could be implemented using the `AutomaticSettings.Trigger.addCustomLoadOverride()` and `AutomaticSettings.Trigger.addCustomSaveOverride()` functions, but that would obviously result in a lot of code duplication between extensions which eliminates many of the benefits of this library. Unlike #13, which mainly applies to Thunderbird, I believe this functionality would also be used by many extensions, including Unicodify and at least three of the prototype extensions I shared with you, so it would be great if it were supported by this library.

> The CSS one.

If I understand you correctly, you are proposing we add a CSS class to each of the buttons. Note that an extension could have multiple options where they need to have more than one instance of an options group and thus they would need multiple tables. This library would then need to know which buttons corresponded to which table. At minimum, the library would need a reference to the table, template for each row, add entry button and remove all entries button.

Anyway, I believe I implemented all your suggested changes to my demo, although I am not sure I understand what you are proposing we update in your library. Note that this issue is requesting support for two similar features:
* Adding more than one instance of an option group.
* Select elements where the `multiple` attribute is set.

My demo was only a possible UI for the former, since you said:
> the UI is totally missing here.
There is not standard HTML component for that.

**Comment by rugk** [comment]
> This library would then need to know which buttons corresponded to which table. At minimum, the library would need a reference to the table, template for each row, add entry button and remove all entries button.

Yes, that's easy though.  One can get it as this `document.querySelector("#TemplatetHIng > .clear-button")`.

----

Yeah as for how exactly to implement it I am also not sure, I'd need to look into the lib too anyway. I did not have a look there since quite some time. :upside_down_face:

**Comment by tdulcet** [comment]
I was finally able to integrate my demo with the `AutomaticSettings.Trigger.addCustomLoadOverride()` and `AutomaticSettings.Trigger.addCustomSaveOverride()` functions to create a working options page that supports more than one instance of an option group:
![image](https://user-images.githubusercontent.com/15305150/177001589-01144cae-170a-4224-b531-bb3bab378106.png)

It actually has five such options that need more than once instance of an options group for each of the five supported issue trackers. Anyway, I found a few issues with this library that I had to work around:
* HTML input elements without the `name` attribute set results in an options group with an `undefined` key. These lines do not first check that the `name` is defined before adding it to the options group: https://github.com/TinyWebEx/AutomaticSettings/blob/fce7d28ce753354da18cad9cfe71c729ec46d424/internal/HtmlModification.js#L155-L159
* This library of course only supports the `input` and `change` events, but it would likely need to use a [MutationObserver](https://developer.mozilla.org/en-US/docs/Web/API/MutationObserver) to detect when the user adds/removes rows from the table or moves them up/down. I was only able to work around this by [manually creating](https://developer.mozilla.org/en-US/docs/Web/Events/Creating_and_triggering_events#triggering_built-in_events) synthetic `change` events.
* The `registerSave()` callback runs before the `addCustomSaveOverride()` callback, so it is not able to take advantage of any fixes to the data made in the latter. It would be very helpful if we could directly override the `HtmlMod.applyOptionToElement()` and  `HtmlMod.getIdAndOptionsFromElement()` functions instead of having to use the `addCustomLoadOverride()` and `addCustomSaveOverride()` callbacks respectively. This would allow developers to fix the values earlier, so the other callbacks would just work as expected.
* The `addCustomLoadOverride()` and `addCustomSaveOverride()` callbacks require a return value. See #22 for more information.
* It would also be helpful if the `addCustomSaveOverride()` function also accepted an individual option, like the `addCustomLoadOverride()` function does, instead of having to handle every option in the options group at once.

I would be happy to share my code if that would be helpful.

**Comment by rugk** [comment]
That all sounds very reasonable and although I don't have that code in mind in detail, I guess yeah… we could change that. Sure sharing your code never hurts (it hopefully lives in some GitHub repo in a branch anyway :wink:). Of course, best would be to split this issue into smaller ones, and fix them in this lib one by one then…

**Comment by tdulcet** [comment]
> Sure sharing your code never hurts (it hopefully lives in some GitHub repo in a branch anyway 😉).

Yes, I had not pushed it yet, but here it is now: https://github.com/tdulcet/Bug-Opener/blob/f250b97540093f103a9a4dcd32c9141f9ec06da0/options/modules/CustomOptionTriggers.js#L19-L120
